### PR TITLE
ref: Use average color for redact masking

### PR DESCRIPTION
**Pull request by brustolin**
Use the text color or the average color of the image as the redaction mask.

_#skip-changelog_

### CODE REVIEW

File: Sources/Sentry/SentrySessionReplay.m [hunk 1]
```diff
@@ -306,11 +306,12 @@ - (void)takeScreenshot
...
-    UIImage *screenshot = [_screenshotProvider imageWithView:_rootView options:_replayOptions];
-
-    _processingScreenshot = NO;
-
-    [self->_replayMaker addFrameAsyncWithImage:screenshot];
+    [_screenshotProvider imageWithView:_rootView
+                               options:_replayOptions
+                            onComplete:^(UIImage *screenshot) {
+                                self->_processingScreenshot = NO;
+                                [self->_replayMaker addFrameAsyncWithImage:screenshot];
```
Comment (philipphofmann) [inline Sources/Sentry/SentrySessionReplay.m:313]: `m`: Can this lead to a retain cycle? Should we maybe use `__weak` to self here?

File: Sources/Swift/Tools/RedactRegion.swift [hunk 2]
```diff
@@ -0,0 +1,48 @@
+#if canImport(UIKit)
+import Foundation
+import UIKit
+
+struct RedactRegion {
+    let rect: CGRect
+    let color: UIColor?
+    
+    init(rect: CGRect, color: UIColor? = nil) {
```
Comment (philipphofmann) [inline Sources/Swift/Tools/RedactRegion.swift:9]: `m`: What made you choose to make the color nullable? Only the tests omit this parameter. The source code always passes in the color.

File: Tests/SentryTests/RedactRegionTests.swift [hunk 3]
```diff
@@ -0,0 +1,102 @@
+import Foundation
+import Nimble
+@testable import Sentry
+import XCTest
+#if os(iOS) || os(tvOS)
+class RedactRegionTests: XCTestCase {
```
Comment (philipphofmann) [inline Tests/SentryTests/RedactRegionTests.swift:6]: `m`: Those are neat test cases. I think it would make sense to add some test cases when the CGRects don't overlap exactly. For example, what happens if the region to subtract is bigger than the sut region.
For example
```
func testSplitBySubtracting_BottomIsWider() {
    let sut = RedactRegion(rect: CGRect(x: 0, y: 0, width: 100, height: 100), color: .red)
    
    let result = sut.splitBySubtracting(region: CGRect(x: 0, y: 50, width: 150, height: 50))
    
    expect(result.count) == 1
    expect(result.first?.rect) == CGRect(x: 0, y: 0, width: 100, height: 50)
    expect(result.first?.color) == .red
}
```

File: Sources/Swift/Tools/RedactRegion.swift [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    }
+    
+    func splitBySubtracting(region: CGRect) -> [RedactRegion] {
+        guard rect.intersects(region) else { return [self] }
+        guard !region.contains(rect) else { return [] }
+        
+        let intersectionRect = rect.intersection(region)
+        var resultRegions: [CGRect] = []
+        
+        // Calculate the top region.
```
Comment (philipphofmann) [inline Sources/Swift/Tools/RedactRegion.swift:21]: Those are useful comments. Thanks 💯

File: Sources/Swift/Tools/SentryViewPhotographer.swift [hunk 5]
```diff
@@ -1,16 +1,12 @@
...
 import CoreGraphics
 import Foundation
 import UIKit
 
-@available(iOS, introduced: 16.0)
-@available(tvOS, introduced: 16.0)
 @objcMembers
-class SentryViewPhotographer: NSObject {
-    
+class SentryViewPhotographer: NSObject, SentryViewScreenshotProvider {
```
Comment (philipphofmann) [inline Sources/Swift/Tools/SentryViewPhotographer.swift:9]: `m`: This class has a low test coverage. Does it make sense to add some tests for these changes in this PR?

File: Sources/Swift/Tools/SentryViewPhotographer.swift [hunk 6]
```diff
@@ -29,23 +25,62 @@ class SentryViewPhotographer: NSObject {
...
+        let redact = self.mask(view: view, options: options)
+        let imageSize = view.bounds.size
+        DispatchQueue.global().async {
+            let screenshot = UIGraphicsImageRenderer(size: imageSize, format: .init(for: .init(displayScale: 1))).image { context in
+                image.draw(at: .zero)
+                
+                for region in redact {
+                    (region.color ?? self.averageColor(of: context.currentImage, at: region.rect)).setFill()
+                    context.fill(region.rect)
+                    //region.image?.draw(in: region.rect)
```
Comment (philipphofmann) [inline Sources/Swift/Tools/SentryViewPhotographer.swift:42]: `l`: Commented code.

File: Sources/Sentry/SentrySessionReplay.m (hunk 1 above)
Comment (brustolin) [reply Sources/Sentry/SentrySessionReplay.m:313]: It can hold the reference for a little longer. But will not create a retain cycle, once the callback returns everything will be disposed.

File: Sources/Swift/Tools/SentryViewPhotographer.swift (hunk 6 above)
Comment (brustolin) [inline Sources/Swift/Tools/SentryViewPhotographer.swift:42]:
```suggestion
```

File: Sources/Swift/Tools/SentryViewPhotographer.swift (hunk 5 above)
Comment (brustolin) [reply Sources/Swift/Tools/SentryViewPhotographer.swift:9]: This bugs me too, but this class relies too much on UI and drawing, testing the Image results will most likely be flaky.

I will try to extract some code from here that can be tested.

File: Sources/Swift/Tools/RedactRegion.swift (hunk 2 above)
Comment (brustolin) [reply Sources/Swift/Tools/RedactRegion.swift:9]: Color can be nil, see `color` function from `SentryViewPhotographer`

```swift
    private func color(for view: UIView) -> UIColor? {
         return (view as? UILabel)?.textColor
     }
```

File: Sources/Sentry/SentrySessionReplay.m (hunk 1 above)
Comment (philipphofmann) [reply Sources/Sentry/SentrySessionReplay.m:313]: But we call onComplete with a strong reference to self async here https://github.com/getsentry/sentry-cocoa/blob/cb3bca31cafa521d1bfe3e3e627d9284d95c7e6e/Sources/Swift/Tools/SentryViewPhotographer.swift#L35-L45

We have to be careful not to introduce retains cycles in the future. Currently, it doesn't matter much, because we always have a strong reference to `SentrySessionReplay` once session replay is enable. Maybe it's safer to use a `__weak` reference now to avoid problems in the future. Up to you.

File: Tests/SentryTests/UIImageHelperTests.swift [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    private let testFrame = CGRect(x: 0, y: 0, width: 100, height: 100)
+    
+    func testAverageColorRed() {
+        let begin = Date()
+            let image = UIGraphicsImageRenderer(size: testFrame.size).image { context in
+                UIColor.red.setFill()
+                context.fill(testFrame)
+            }
+            
+            expect(UIImageHelper.averageColor(of: image, at: self.testFrame)) == .red
```
Comment (philipphofmann) [inline Tests/SentryTests/UIImageHelperTests.swift:18]: `l`:  The code formatting here seems to be a bit off.

File: Tests/SentryTests/UIImageHelperTests.swift [hunk 8]
```diff
@@ -0,0 +1,65 @@
+#if canImport(UIKit)
+import Foundation
+import Nimble
+@testable import Sentry
+import XCTest
+
+class UIImageHelperTests: XCTestCase {
```
Comment (philipphofmann) [inline Tests/SentryTests/UIImageHelperTests.swift:7]: Neat tests 👏

File: Tests/SentryTests/RedactRegionTests.swift [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+        
+        let result = sut.splitBySubtracting(region: CGRect(x: 0, y: 25, width: 100, height: 50))
+        
+        expect(result.count) == 3
+        expect(result[0].rect) == CGRect(x: 50, y: 0, width: 100, height: 25)
+        expect(result[1].rect) == CGRect(x: 50, y: 75, width: 100, height: 25)
+        expect(result[2].rect) == CGRect(x: 100, y: 25, width: 50, height: 50)
+    }
+
+    func testSplitBySubtracting_TopIsWider() {
```
Comment (philipphofmann) [inline Tests/SentryTests/RedactRegionTests.swift:101]: Thanks for adding more tests for these edge cases 💯

File: Tests/SentryTests/UIRedactBuilderTests.swift [hunk 10]
```diff
@@ -0,0 +1,130 @@
+#if os(iOS)
+import Foundation
+import Nimble
+@testable import Sentry
+import UIKit
+import XCTest
+
+class UIRedactBuilderTests: XCTestCase {
```
Comment (philipphofmann) [inline Tests/SentryTests/UIRedactBuilderTests.swift:8]: Those tests are the specification to the sut. I don't need to read the code of the sut to understand what the class does. Excellent 🥇

File: Tests/SentryTests/UIRedactBuilderTests.swift [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+    
+    private class RedactOptions: SentryRedactOptions {
+        var redactAllText: Bool = true
+        var redactAllImages: Bool = true
+    }
```
Comment (philipphofmann) [inline Tests/SentryTests/UIRedactBuilderTests.swift:13]: `m`: If I'm not mistaken we always test with both values true. It would be great to test with different combinations of these too.

File: Sources/Swift/Tools/UIRedactBuilder.swift [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+    //This is a list of UIView subclasses that will be ignored during redact process
+    var ignoreClasses: [AnyClass] 
+    //This is a list of UIView subclasses that need to be redacted from screenshot
+    var redactClasses: [AnyClass]
+    
+    init() {
+        redactClasses = [ UILabel.self, UITextView.self, UITextField.self ] + [
+            "_TtCOCV7SwiftUI11DisplayList11ViewUpdater8Platform13CGDrawingView",
+            "_TtC7SwiftUIP33_A34643117F00277B93DEBAB70EC0697122_UIShapeHitTestingView",
+            "SwiftUI._UIGraphicsView", "SwiftUI.ImageLayer"
```
Comment (philipphofmann) [inline Sources/Swift/Tools/UIRedactBuilder.swift:61]: `m`: Where do these class names come from? Can you add a comment to explain this?

File: Sources/Swift/Tools/UIRedactBuilder.swift [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+    
+    private func shouldRedact(view: UIView, redactText: Bool, redactImage: Bool) -> Bool {
+        if redactImage, let imageView = view as? UIImageView {
+            return shouldRedact(imageView: imageView)
+        }
+        return redactText && redactClasses.contains { view.isKind(of: $0) }
+    }
+    
+    private func shouldRedact(imageView: UIImageView) -> Bool {
+        // Checking the size is to avoid redact gradient backgroud that
```
Comment (philipphofmann) [inline Sources/Swift/Tools/UIRedactBuilder.swift:95]: `l`
```suggestion
        // Checking the size is to avoid redact gradient background that
```

File: Sources/Swift/Tools/UIRedactBuilder.swift [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+            }
+        }
+    }
+    
+    private func color(for view: UIView) -> UIColor? {
+        return (view as? UILabel)?.textColor
+    }
+    
+    private func hasBackground(_ view: UIView) -> Bool {
+        return view.backgroundColor != nil && (view.backgroundColor?.cgColor.alpha ?? 0) > 0.9
```
Comment (philipphofmann) [inline Sources/Swift/Tools/UIRedactBuilder.swift:131]: `m`: Why only if the alpha is greater than `0.9`? Can you add this info as a comment?

File: Sources/Swift/Tools/UIRedactBuilder.swift [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+            return shouldRedact(imageView: imageView)
+        }
+        return redactText && redactClasses.contains { view.isKind(of: $0) }
+    }
+    
+    private func shouldRedact(imageView: UIImageView) -> Bool {
+        // Checking the size is to avoid redact gradient backgroud that
+        // are usually small lines repeating
+        guard let image = imageView.image, image.size.width > 10 && image.size.height > 10  else { return false }
+        return image.imageAsset?.value(forKey: "_containingBundle") == nil
```
Comment (philipphofmann) [inline Sources/Swift/Tools/UIRedactBuilder.swift:98]: `l`: Out of curiosity. What is the `_containingBundle`?

**Review by philipphofmann — COMMENTED** [review]
Thanks for adding more tests and extracting some code to make it testable 👍

File: Sources/Swift/Tools/UIImageHelper.swift [hunk 16]
```diff
@@ -0,0 +1,36 @@
+#if canImport(UIKit) && !SENTRY_NO_UIKIT
+#if os(iOS) || os(tvOS)
+
+import Foundation
+import UIKit
+
+final class UIImageHelper {
+    private init() { }
+    
+    static func averageColor(of image: UIImage, at region: CGRect) -> UIColor {
```
[the remark this answers is not in the document]
Comment (brustolin) [reply Sources/Swift/Tools/UIImageHelper.swift:10]: I prefer to organise helper functions into static classes. 
We dont have static class in Swift, so static function in a class that cannot be initialised it is.

File: Tests/SentryTests/UIRedactBuilderTests.swift (hunk 11 above)
Comment (brustolin) [reply Tests/SentryTests/UIRedactBuilderTests.swift:13]: True. Forgot to test no redacting when the flag is false

File: Sources/Swift/Tools/UIRedactBuilder.swift (hunk 15 above)
Comment (brustolin) [reply Sources/Swift/Tools/UIRedactBuilder.swift:98]: Information about the bundle that the image is coming from.

**Review by philipphofmann — APPROVED** [review]
LGTM. Thanks for including all the feedback.

**Comment by bruno-garcia** [comment]
hope we can ship this one Monday to unblock customers trialing who are affected by high CPU
